### PR TITLE
Circular dependencies

### PR DESCRIPTION
**Pull request by mthmulders**
Based on discussions in #2990, here's a draft PR that implements circular dependencies. Any feedback is welcome.

### CODE REVIEW

**Comment by manovotn** [comment]
I'll review this today/tomorrow and provide some feedback.
There are some test failures though - NPE from Arc processor, coming only from agroal tests though.

**Comment by mthmulders** [comment]
Yep, I noticed those failures. I'm trying to debug them, see what I can do to fix those.

**Review by manovotn — COMMENTED** [review]
This is looking really nice overall, I do have a few questions/remarks (some in code, some here):
* Do we really need to apply the suppliers to interceptor providers as well? Those shouldn't be part of any cycle.
* Disposer providers - I know we need to replace those, but it's not nice to see another suppliers in play. These would deserve a separate test by the way.
* Built-in beans - why do you need suppliers there? I cannot imagine a situation where you need supplier for bean manager?

Now we need to think of a way to only use suppliers for bean that we know are a part of a circular dependency, so that, ideally, an app with no such dependencies won't have to use any extra suppliers. But if I am not mistaken, we currently generate `_Bean` classes prior to the check for circularity. If we do the check earlier, we should be able to keep an information on the circular deps and then create bean instance with that in mind?
Not that I would know how to do it from the top of my head :-) Just thinking aloud.

**Comment by manovotn** [comment]
I also added Martin for review but he won't be looking at it sooner then next week.
And don't mind the Windows CI failure, that one is unrelated.

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/circular/CircularDependenciesTest.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import static org.junit.Assert.assertThat;
+
+import io.quarkus.arc.test.ArcTestContainer;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.spi.CDI;
+import javax.inject.Inject;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class CircularDependenciesTest {
```
[the remark this answers is not in the document]
Comment (mthmulders) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/circular/CircularDependenciesTest.java]: Of course we can. Consider it done.

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/injection/erroneous/CircularInjectionNotSupportedTest.java [hunk 2]
```diff
@@ -21,8 +25,10 @@ public Statement apply(Statement base, Description description) {
                 public void evaluate() throws Throwable {
                     try {
                         base.evaluate();
+                        fail("Expected an IllegalStateException to be thrown, but it wasn't");
                     } catch (IllegalStateException e) {
                         // expected failure on ISE due to circular dependency
+                        assertThat(e.getMessage(), containsString("Circular dependencies not supported"));
```
[the remark this answers is not in the document]
Comment (mthmulders) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/injection/erroneous/CircularInjectionNotSupportedTest.java]: Fair enough. Agreed that we keep the `fail` when the ISE isn't thrown?

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/circular/CircularDependenciesChainTest.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import static org.junit.Assert.assertThat;
+
+import io.quarkus.arc.test.ArcTestContainer;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.spi.CDI;
+import javax.inject.Inject;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class CircularDependenciesChainTest {
```
[the remark this answers is not in the document]
Comment (mthmulders) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/circular/CircularDependenciesChainTest.java]: Please do! The more of these test cases we have, the better it is, I'd say.

**Comment by mthmulders** [comment]
Thanks for this first round of review!

Also thinking out loud... if we have Suppliers only when we really need them (i.e., in case of circular dependencies), our generated code will become full of

    if (circularDependency2) {
        injectProviderSupplier2.get().whatever();
    } else {
        injectProvider2.whatever();
    }

I'm afraid this will be hard to understand and hard to maintain, won't it?

Also thinking out loud (and not sure if it's actually possible)... What if the generated `_Bean` classes would have access to the full `Map<String, InjectableBean<?>> beanIdToBean` that we currently keep in the `ComponentsGenerator`, so they can lookup their dependencies at runtime (based on the bean ID)?

**Comment by mthmulders** [comment]
@manovotn, what do you think of the idea to give the generated `_Bean` classes access to the full `Map<String, InjectableBean<?>> beanIdToBean`?

@mkouba Did you have a chance already to take a look at this?

**Comment by mkouba** [comment]
> what do you think of the idea to give the generated _Bean classes access to the full Map<String, InjectableBean<?>> beanIdToBean?

That's an interesting idea. 

However, a generated "lambda" supplier class looks like this:
```java
public class CircularDependenciesChainTest_ComponentsProvider$$function$$576 implements Supplier {
	private final HashMap f0;

	@Override
	public Object get() {
		return this.f0.get("f6b61e7996d82c3ad15aafa7f7f3f3be51915e29");
	}

	public CircularDependenciesChainTest_ComponentsProvider$$function$$576(final HashMap f0) {
		this.f0 = f0;
	}
}
```
So we could get rid of these lambdas having a class like this:
```java
public class BeanSupplier implements Supplier {
        private final HashMap map;
        private final String id;

       public BeanSupplier(HashMap map, String id) {
         this.map = map;
         this.id = id;
       }

	@Override
	public Object get() {
		return map.get(id);
	}
}
```
And use instances of this class instead. So we would replace the `new CircularDependenciesChainTest_ComponentsProvider$$function$$576(map)` in the `ComponentsProvider` with something like `new BeanSupplier(map,"372044bb8364ed1d4d4923339ba742dd00c73713")`. WDYT?

I'll let @manovotn to review the rest of the PR... (once he's back from PTO ;-).

**Comment by mthmulders** [comment]
Thanks @mkouba for the idea. It has flewn through my head but I rejected it as not feasible, but since you thought of it too I gave it a try. It *is* feasible! I've implemented two variants, one with the  `Map` and the pre-configured ID, and one where we know in advance what the value is that must be returned. Of course, naming is open for discussion.

The good news is, it brings down the number of generated lambda in the arc/tests module. Before I started, I counted 41. My initial implementation went to 708, and now we're back at 41. So with this PR we will not create a single extra lambda instance.

**Comment by mkouba** [comment]
> My initial implementation went to 708, and now we're back at 41.

+1!

@mthmulders Could you  pls rebase?

**Comment by mthmulders** [comment]
> > My initial implementation went to 708, and now we're back at 41.
> 
> +1!
> 
> @mthmulders Could you pls rebase?

Done, and did a force push. On my local machine, there's a failure in the Kotlin-integration. It seems unrelated to my changes but I'm not 100% sure. Also, I'm not 100% sure if I merged the changes in 4b147c83 correctly. Would appreciate if you could double-check.

**Review by manovotn — APPROVED** [review]
The idea with Map/Fixed suppliers is really good!
All tests are passing, IMO this PR is good to be merged.

**Review by mkouba — APPROVED** [review]

**Pull request merged**